### PR TITLE
Use a special matrix.to URL for room notifs

### PR DESCRIPTION
**Pull request by dbkr**
So you'll have to explicitly accept the pill completion to actually
do a room notification.

Note that this matches on formatted_body as the URL will only be
in the formatted_body. It could also match for only HTML bodies,
but the URL ought to be specific enough that it's sensible to match
whatever the format is?

### CODE REVIEW

**Comment by maxidor** [comment]
@dbkr Why not use a URI with a dedicated protocol instead of a URL which is linked to some ownership?

**Comment by erikjohnston** [comment]
So I have a number of questions/comments:

1. Do we want it to bing when someone posts the link? Or should we actually be checking its a pill?
2. We've so far avoided putting matrix.to into the spec, no? At least afaik its not mentioned anywhere in the server. What will the link even show? TBH I'm confused why matrix.to is involved at all.
3. Does this mean that we're moving towards requiring HTML support for "usable" clients?

**Comment by ara4n** [comment]
@maxidor:
 1. so that clients which don't understand the feature can view something sensible that explains what's going on and what they're missing out on.
 2. it just requires that the client can parse HTML to find the link (as opposed to render it) - which does not seem entirely unreasonable.

@erikjohnston:
 1. i'd assume we'd want to check that it's in a pill
 2. matrix.to is being used here as a namespace (like xmlns), whilst providing fallback to an explanatory page.
 3. no, this doesn't require "HTML support", but yes, the proposal means you'd have to parse the HTML to find the existence of the bing.

Perhaps using metadata on the event is going to be an easier fix (and the pill then ends up being entirely cosmetic)?

**Comment by maxidor** [comment]
> it just requires that the client can parse HTML to find the link (as opposed to render it) - which does not seem entirely unreasonable.

Few future directions that could make this a bad idea:
- Homeserver and/or Bridges with content restrictions like spam protection (no links, or only allowed domains)
- @.... is the start of Matrix IDs, which could break clients or trigger when not wanted
- Low bandwith optimization where all non-necessary are taken out (only text content is mandatory as per spec)

> Perhaps using metadata on the event is going to be an easier fix (and the pill then ends up being entirely cosmetic)?
What about only make it metadata and let the client create the pill itself if wanted/needed by the user?

**Comment by ara4n** [comment]
Just had an IRL chat with @dbkr & @erikjohnston.  Conclusions were:
 * All agreed that requiring parsing HTML is horrid.
 * We could put metadata on the event (something like `notify: room`), but this will be quite painful to migrate to - mobile clients won't be able to send @-room's until it lands, for instance.
 * So as a quick fix, let's just prompt on the clients before sending messages that include `@room`, and just match based on the blunt string.  The client can put a pill in the HTML linking to https://matrix.to/@room (*purely as a namespace and for compatibility with dumb clients which can do HTML but not the `@room` feature*) around `@room` to make it prettier and easier to highlight.

In future we could migrate to a the `notify` metadata, which could potentially also be used to handle high-priority notifications for messages and features like that.

**Comment by maxidor** [comment]
@ara4n does that mean the text version of the message keeps `@room` in it, so text-only clients don't need to parse HTML?

**Comment by dbkr** [comment]
Sounds good - closing this for now then

**Comment by ara4n** [comment]
@maxidor: yes.

**Comment by turt2live** [comment]
fwiw, I have some thoughts on the riot-side bug: https://github.com/vector-im/riot-web/issues/5420

tldr is `mentions: ['@room']` to support the possibility of pushing user mentions into there as well.
